### PR TITLE
Improvements to Platforms Documentation

### PR DESCRIPTION
**Pull request by ben-albrecht**
- List platforms into categories

### CODE REVIEW

**Comment by bradcray** [comment]
I like listing the platforms in priority (maturity?), rather than alphabetical, order.

I'm not as crazy about the title changes.  E.g., I think the content of the "Platforms" page reads well, but a document entitled "Cray Systems" suggests to me that the document is going to tell me about Cray systems when in fact it's describing the use of Chapel on Cray systems (so I liked the old title better when viewing that page itself).  Similarly, the new title for the "Platforms" page suggests that these are the only platforms that Chapel runs on where I felt like the old title implied a bit better that these are the platforms that have/require/benefit from additional documentation, while a platform not listed here might work just fine and need no special documentation (e.g., linux64).

**Comment by mppf** [comment]
I like this table from the libuv project describing how much each platform is supported:

 https://github.com/libuv/libuv/blob/master/SUPPORTED_PLATFORMS.md

**Comment by ben-albrecht** [comment]
@mppf -- me too. We could include the links to the specific documentation for platforms with platform notes as well. Curious on @bradcray's thoughts though.

**Comment by bradcray** [comment]
Looks like a beast to maintain (particularly the "which versions" fields).  Again, the goal of this page has traditionally not been to provide a list of which platforms we support (as the new title suggests to me), but rather as a place to store platform-specific notes for those cases where the generic instructions are insufficient or incomplete.

If we were to do something like this table, I'd use some sort of visual icons rather than words to represent different levels of support using qualitative notions like:
* tested nightly / regularly
* tested occasionally
* users have reported success running on this platform

But really, I think creating such a table is a time trap.  It will only become dated/incorrect over time without tons of testing that we don't do (and, I'd argue, don't want to do.  I couldn't even get an answer to my "should we test all the OSes that Michael manually tested before the last release?" let alone start to think about opening it up to sweeping different OS versions or compiler versions to see where the breaking points are...).  It's also not clear to me that the lack of this table is hurting us in any way.

This PR, like a lot of stuff lately, begs for me the question "Why are we working on this rather than the priorities we've identified?"  Which provides part of the answer as to why I never get any actual work done...

**Comment by ben-albrecht** [comment]
Noted. We can revisit this in the future.

**Comment by bradcray** [comment]
Again, I do like re-listing the platforms in priority (maturity?), rather than alphabetical, order and would merge that commit, esp. having spent the time to review this PR.

Thinking more about a more general "supported platforms" listing on the way in, here are some further thoughts to explain my reaction:  Let's say we did list supported platforms and even versions.  In addition to the maintenance overhead and sense that this hasn't been holding us back in any way, I think it actually does a dis-service to the project by complicating the information presented to the user and potentially having users prematurely discount themselves as being a supported platform or of Chapel being likely to work for them.  I.e., "I don't see my platform here, so I won't even bother trying Chapel."  In contrast, my strong preference would be to continue the simpler message we say today (in talks at least, though it'd be great to reflect this in the docs as well):  That Chapel should run on any platform that has:

* a UNIX-like environment (e.g., bash, [g]make, python, ...)
* a C and C++ compiler
* POSIX threads
* (for distributed memory execution) some way of communicating like UDP, MPI, RDMA

Potentially backed up by the prerequisites file for more details (e.g., supported Python versions).  This makes a strong statement that "We should run on your platform!" and ought to maximize the number of people who give it a try.  If we then also make it clear that they should let us know if things don't work for them, it will help us find portability issues (or improve the prerequisites list).  Whereas if we list "supported platforms" we create work for ourselves and cause people to self-select out prematurely.  I'm thinking specifically of the guy at SC who ran Chapel on a low-power system that I've never even heard of and the fact that he bothered to give it a try and had success.  I don't want us to be in the business of pre-determining whether such cases will work nor closing them off by having a long list of supported platforms.

**Comment by ben-albrecht** [comment]
@bradcray Change made - let me know if it passes review, and I'll merge.

I think the platforms page could eventually be a combination of what has been discussed here, but again - let's revisit in the future.

e.g.,

```

Platforms
=========

Chapel should run on any platform that has:

    - a UNIX-like environment (e.g., bash, [g]make, python, ...)
    - a C and C++ compiler
    - POSIX threads
    - (for distributed memory execution) some way of communicating like UDP, MPI, RDMA

[Table of some platforms that meet this criteria, 
 with links to platform-specific notes, if available, 
 and denoting if it's tested (and maybe for what versions it is tested)]
```

File: doc/platforms/index.rst [hunk 1]
```diff
@@ -1,15 +1,17 @@
 .. _platforms-index:
 
-Platform-Specific Notes
-=======================
-
-Contents:
-
-.. toctree::
-   :hidden:
+Platforms-Specific Notes
```
Comment (bradcray) [inline doc/platforms/index.rst]: The 's' in Platforms doesn't parse for me.  Is it meant to be possessive?  If plural, I'd think that "Notes" would be all that would need to be pluralized, not "Platform-specific" (which I think serves as an adjective -- "what kind of notes"?)

Comment (ben-albrecht) [inline doc/platforms/index.rst]: Typo'd. Intended to be "Platform-Specific Notes"

File: doc/platforms/index.rst [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
-   **
+   cray
+   macosx
+   cygwin
+   knl
+   infiniband
+   udp
+   ibm
+   marenostrum
+   sgi
```
Comment (bradcray) [inline doc/platforms/index.rst]: Here's where I realize and have to confess that I didn't actually look at the order the first time through, just liked the idea of sorting based on priority.  I might be tempted to reorder to the following, primarily to take cray out of the #1 slot and emphasize the notes that are likely to be useful to a broader group of people:

* macosx (probably most common of the important cases here)
* cray (important to us)
* udp (most common network)
* infiniband (most common network)
* knl
* cygwin
* ...

Or, we could add additional categories which might make the lists a bit easier to visually parse.  For example:

major platforms
----------------
* macosx
* cray
* knl

networks
---------
* udp
* infiniband

older platforms (may be out of date)
-------------------------------------
* ibm
* marenostrum
* sgi

**Comment by bradcray** [comment]
w.r.t. the longer-term vision, I'd probably roll the characterization of what's required into the prereqs doc, as that seems like the most natural place for it (or, put another way, putting it here seems to duplicate what ought to be in the prereqs doc to me).

I'm not sure what "tables of platforms that meet this criteria" adds.  Isn't that effectively "almost everything"?  And again, what does it mean for a platform not to be on there?  (Note also that even the current list are not siblings:  Some are architectures, some are processor types, some are network types, some are specific machines.  I don't have a problem with that per se, but I don't see that there's a single table that would do everything you're proposing here.

Maintaining version numbers of what we test against seems like a thankless maintenance task (someone updated some piece of software so we need to update our documentation), and something that would be better exposed through/by Jenkins than the documentation to me.  I say this having attempted to keep such documentation up-to-date in the past and hating the cost involved (and the fact that it was usually wrong) without seeing much benefit...  I think the perceived benefit is "information!" but given that it's quite possibly incorrect information, that benefit goes away.

**Pull request merged**